### ISSUE TITLE
`plot_matrix` colour bar overlaps with the matrix when plotting on assigned subplot axes

### ISSUE THREAD

**Issue by htwangtw** [nilearn/nilearn#3378]
The `plot_matrix` will overlap the colour bar on the main figure when plotting on subplots in a horizontal layout. This cannot be fixed by adjusting spacing with `subplots_adjust`.

Nilearn version: 0.9.1

### Expected behavior

Colour bar stay in each subplot.

### Actual behavior

Colour bar at weird places.

### Steps and code to reproduce bug

```python
from nilearn import datasets
from nilearn.maskers import NiftiMapsMasker
from nilearn.connectome import ConnectivityMeasure
from nilearn.plotting import plot_matrix

import numpy as np

import matplotlib.pyplot as plt


atlas = datasets.fetch_atlas_msdl()
atlas_filename = atlas['maps']
labels = atlas['labels']

data = datasets.fetch_development_fmri(n_subjects=1)
fmri_filenames = data.func[0]

masker = NiftiMapsMasker(maps_img=atlas_filename, standardize=True, detrend=True)
correlation_measure = ConnectivityMeasure(kind='correlation')

time_series = masker.fit_transform(fmri_filenames)
correlation_matrix = correlation_measure.fit_transform([time_series])[0]
np.fill_diagonal(correlation_matrix, 0)

# plot on subplot, with colorbar
fig = plt.figure()
axes = fig.subplots(1, 2)
for corr, ax in zip([correlation_matrix, correlation_matrix], axes):
    plot_matrix(corr, labels=labels, colorbar=True, axes=ax, vmax=0.7, vmin=-0.7)
```
Output:
![image](https://user-images.githubusercontent.com/13743617/194398382-2905a23e-126d-4ce8-af39-0761291684ec.png)

**Comment by emdupre** [comment]
Can you confirm if the MPL version generating these plots is 3.6.0, @htwangtw, or another version ?

**Comment by alexisthual** [comment]
I observed the same problem when using `plotting.plot_surf()` with subplots (MPL 3.6.0).
In this particular case, I used a `GridSpec` but it doesn't seem to change anything.

![sub-04_sub-05_features_correlation](https://user-images.githubusercontent.com/13835654/194515278-62fa2ea0-3c5c-4fbe-a072-d8d0f39ac19e.jpg)

**Comment by htwangtw** [comment]
> Can you confirm if the MPL version generating these plots is 3.6.0, @htwangtw, or another version ?

I was testing under 3.4.2

**Comment by jeromedockes** [comment]
here is a minimal example for anyone wanting to reproduce the plotting issue without computing the brain connectivity stuff

```python
from matplotlib import pyplot as plt
import numpy as np

from nilearn import plotting

matrix = np.zeros((10, 10))
fig, axes = plt.subplots(1, 2)
for ax in axes:
    plotting.plot_matrix(matrix, axes=ax)

plt.show()
```

![nilearn-plot](https://user-images.githubusercontent.com/9196501/194590574-6e22161e-227a-4bd8-abb2-5c1ee7d5e6be.png)

**Comment by jeromedockes** [comment]
related:
- https://github.com/nilearn/nilearn/issues/545

- also about giving more access to nilearn display objects (for all but the simplest layouts I find it easier to create all the axes, including the colorbars, myself with matplotlib and just use nilearn to plot one brain in one matplotlib Axes at a time):
  - https://github.com/nilearn/nilearn/pull/3073
  - https://github.com/nilearn/nilearn/pull/2859

**Comment by RaphaelMeudec** [comment]
I started looking into it, fixing the weird behaviour, but I'm still investigating on the colorbar labels overflowing on the next subplot.

![image](https://user-images.githubusercontent.com/12402673/198005776-b1c0063f-86f4-42ec-aab0-70e6e890cc19.png)


@alexisthual I don't think I touched code related to surface plot. Could you provide a minimal example so I can see how it relates?

**Comment by alexisthual** [comment]
TLDR: I think `plt.tight_layout()`

However, because I was using nested grids, I wasn't sure who the culprit was :detective: 

Here is a working piece of code:

```python
# %%
import matplotlib.pyplot as plt
from nilearn import datasets, plotting, surface

# %%
motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]
fsaverage = datasets.fetch_surf_fsaverage()
texture = surface.vol_to_surf(stat_img, fsaverage.pial_right)

# %%
n_subplots = 3
n_fig_per_subplot = 2

fig = plt.figure(
    figsize=(3 * n_subplots * n_fig_per_subplot, 3)
)
fig.suptitle("Test suptitle")
outer_grid = fig.add_gridspec(
    1, n_subplots, wspace=0, hspace=0
)

for i in range(n_subplots):
    inner_grid = outer_grid[i].subgridspec(1, n_fig_per_subplot, wspace=0, hspace=0)
    axes = inner_grid.subplots(subplot_kw={"projection": "3d"})

    for j in range(n_fig_per_subplot):
        plotting.plot_surf(
            fsaverage.pial_left,
            texture,
            bg_map=fsaverage.sulc_left,
            bg_on_data=True,
            axes=axes[j],
            cmap="bwr",
            colorbar=True,
        )

# plt.tight_layout()
plt.show()
```

Without `plt.tight_layout()`:
![image](https://user-images.githubusercontent.com/13835654/198408272-da78b21b-ff4d-4f60-87f2-ffbaaa9cc1fb.png)

With `plt.tight_layout()`:
```
<ipython-input-25-115fb66b21bc>:30: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  plt.tight_layout()
```
![image](https://user-images.githubusercontent.com/13835654/198408220-431545f0-db23-4bd8-8b9e-6d4f5666c0f2.png)
